src/deviation.rs

```diff
@@ -0,0 +1,378 @@
+    ///
+    /// **Panics** if the type cast from `A` to `f64` fails.
+    ///
+    /// [root-mean-square error]: https://en.wikipedia.org/wiki/Root-mean-square_deviation
+    fn root_mean_sq_err(&self, other: &ArrayBase<S, D>) -> Result<f64, MultiInputError>
+    where
+        A: AddAssign + Clone + Signed + ToPrimitive;
+
+    /// Computes the [peak signal-to-noise ratio] between `self` and `other`.
+    ///
+    /// ```text
+    /// 10 * log10(maxv^2 / mse(a, b))
+    /// ```
+    ///
+    /// where `self` is `a`, `other` is `b`, `mse` is the mean-squared-error
+    /// and `maxv` is the maximum possible value either array can take.
+    ///
+    /// The following **errors** may be returned:
+    ///
+    /// * `MultiInputError::EmptyInput` if `self` is empty
+    /// * `ShapeMismatch` if `self` and `other` don't have the same shape
+    ///
+    /// **Panics** if the type cast from `A` to `f64` fails.
+    ///
+    /// [peak signal-to-noise ratio]: https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio
+    fn peak_signal_to_noise_ratio(
+        &self,
+        other: &ArrayBase<S, D>,
+        maxv: A,
+    ) -> Result<f64, MultiInputError>
+    where
+        A: AddAssign + Clone + Signed + ToPrimitive;
+
+    private_decl! {}
+}
+
+macro_rules! return_err_if_empty {
+    ($arr:expr) => {
+        if $arr.len() == 0 {
+            return Err(MultiInputError::EmptyInput);
+        }
+    };
+}
+macro_rules! return_err_unless_same_shape {
+    ($arr_a:expr, $arr_b:expr) => {
+        if $arr_a.shape() != $arr_b.shape() {
+            return Err(ShapeMismatch {
+                first_shape: $arr_a.shape().to_vec(),
+                second_shape: $arr_b.shape().to_vec(),
+            }
+            .into());
+        }
+    };
+}
+
+impl<A, S, D> DeviationExt<A, S, D> for ArrayBase<S, D>
+where
+    S: Data<Elem = A>,
+    D: Dimension,
+{
+    fn count_eq(&self, other: &ArrayBase<S, D>) -> Result<usize, MultiInputError>
+    where
+        A: PartialEq,
+    {
+        return_err_if_empty!(self);
+        return_err_unless_same_shape!(self, other);
+
+        let mut count = 0;
+
+        Zip::from(self).and(other).apply(|a, b| {
+            if a == b {
+                count += 1;
+            }
+        });
+
+        Ok(count)
+    }
+
+    fn count_neq(&self, other: &ArrayBase<S, D>) -> Result<usize, MultiInputError>
+    where
+        A: PartialEq,
+    {
+        self.count_eq(other).map(|n_eq| self.len() - n_eq)
+    }
+
+    fn sq_l2_dist(&self, other: &ArrayBase<S, D>) -> Result<A, MultiInputError>
+    where
+        A: AddAssign + Clone + Signed,
+    {
+        return_err_if_empty!(self);
+        return_err_unless_same_shape!(self, other);
+
+        let mut result = A::zero();
+
+        Zip::from(self).and(other).apply(|self_i, other_i| {
+            let (a, b) = (self_i.clone(), other_i.clone());
+            let abs_diff = (a - b).abs();
+            result += abs_diff.clone() * abs_diff;
```

Comment:
Curiosity, more than anything else: how does this compile without a `A: Mul` bound?

Reply:
It's due to the trait inheritance chain in `Signed`:

```
Mul <- NumOps <- Num <- Signed
```

Reply:
Unexpected, but useful.